### PR TITLE
Enable visco-elasticity with two-phase melt transport

### PR DESCRIPTION
**Pull request by ryanstoner1**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

Added the elastic force to the melt equations following [Keller et al.](https://doi.org/10.1093/gji/ggt306) (2013) so that the matrix can be visco-elasto-plastic even with porous flow. 

The primary things to verify are in melt.cc. 

1. I use `p_c_scale` if there is no porosity to enforce no divergence. This was already done, but it was applied to an additional term and helped with convergence of the bending beam. There might be a better way to consolidate. 

2. Elasticity has to use discontinuous elements, so I want to make sure that is okay correctness-wise. I relaxed the guard for those. 

3. Keller et al. use an exponent melt fraction-dependence q is used to scale the reference pore modulus. For now it's left hardcoded.

4. I tried to test in melt.cc whether a timestep is properly defined past initialization with `dt > 0.0`. Not sure if there's a better way.

Right now I have the viscoelastic_beam_benchmark at 0 melt limit and an elastic relaxation timescale check. 
Any feedback would be much appreciated!
(@naliboff @tjhei @jdannberg) 

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

AI disclosure note: a draft of .prm and melt changes used an LLM. All such changes were checked and/or rewritten by me, and tests were re-run before submitting.

### CODE REVIEW

File: doc/modules/changes/20260714_stoner [hunk 1]
```diff
@@ -0,0 +1,15 @@
+New: Elasticity can now be combined with two-phase melt transport. The melt
+Stokes assembler adds the stored elastic (deviatoric) shear stress as a force
+on the solid momentum equation, and can optionally treat the compaction of the
+solid matrix as visco-elastic via a new "Reference pore modulus" parameter in
+the "Melt settings" subsection (following Keller, May & Kaus, 2013, who define
+the porosity-dependent elastic pore modulus K_phi = K_0 phi^(-1/2)); the default
+value recovers purely viscous compaction, so existing models are unchanged, so in cells without melt the melt momentum operator reduces exactly
```
Comment (naliboff) [inline doc/modules/changes/20260714_stoner]: Perhaps break this line up into two for uniform formatting?

File: source/simulator/melt.cc [hunk 2]
```diff
@@ -277,7 +277,25 @@ namespace aspect
...
+          // The elastic pore modulus is K_phi = K_0 * phi^(-q) (q = 0.5, Keller, May &
+          // Kaus 2013, eq. 52), with K_0 the reference pore modulus.
+          // Set viscoelastic viscosity from elastic and viscous contributions
+          // (eq. 29).
+          const double dt = this->get_timestep();
+          const double K_0_pre = this->get_melt_handler().melt_parameters.reference_pore_modulus;
+          const unsigned int por_idx = introspection.compositional_index_for_name("porosity");
+          const double porosity = std::max(scratch.material_model_inputs.composition[q][por_idx], 0.0);
+          double viscosity_c_ve = viscosity_c;
+          if (dt > 0.0 && K_0_pre < 1e30 && porosity > 0.0)
```
Comment (naliboff) [inline source/simulator/melt.cc]: I don't have a better suggestion at the moment, but I wonder if there is a more elegant way to handle the cutoff for ` K_0_pre`?

File: source/simulator/melt.cc [hunk 3]
```diff
@@ -542,11 +589,25 @@ namespace aspect
...
               if (scratch.rebuild_stokes_matrix)
                 for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
                   {
                     data.local_matrix(i,j) += ( (eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
-                                                - (eta_two_thirds * (scratch.div_phi_u[i] * scratch.div_phi_u[j])
+                                                // The -2/3 eta (div u)(div v) deviatoric/compaction term only
+                                                // belongs in cells where the solid actually compacts (melt cells).
+                                                // In non-melt cells (p_c pinned, phi->0) the solid is incompressible,
+                                                // so gate it out there to recover the exact incompressible operator.
+                                                - ( (p_c_scale > 0.0 ? 1.0 : 0.0)
```
Comment (naliboff) [inline source/simulator/melt.cc]: Will be interesting to see if this leads to tests breaking. Can you add a note here explaining why this is needed when elasticity is added? I still don't fully understand why the benchmarks with no melt present break when elasticity is added (i.e., why are the regular assemblers not being invoked)?

File: source/simulator/melt.cc [hunk 4]
```diff
@@ -1839,12 +1935,31 @@ namespace aspect
...
     // to the DG formulation:
-    AssertThrow(!this->get_parameters().use_discontinuous_temperature_discretization &&
-                !this->get_parameters().have_discontinuous_composition_discretization,
-                ExcMessage("Using discontinuous elements for temperature "
-                           "or composition in models with melt transport is currently not implemented.") );
+    AssertThrow(!this->get_parameters().use_discontinuous_temperature_discretization,
+                ExcMessage("Using discontinuous elements for temperature in models "
+                           "with melt transport is currently not implemented.") );
+
+    // Discontinuous elements for the compositional fields are likewise not
```
Comment (naliboff) [inline source/simulator/melt.cc]: I will come back to this tomorrow, but can we in fact run models with fields for both elasticity stresses and melt? My recollection was that this would produce issues with the operator splitting routines?

**Review by naliboff — COMMENTED** [review]
@ryanstoner1 - Thank you for putting all of this together! I just did an initial scan through the PR and made a few minor comments. I will review this in more detail first thing tomorrow with a fresh set of eyes.

File: tests/melt_bending_beam_elasticity.prm [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# Viscoelastic bending beam with two-phase melt transport at zero porosity.
```
Comment (naliboff) [inline tests/melt_bending_beam_elasticity.prm]: Maybe add an additional line stating that the results should be identical to the regular test case for the zero porosity scenario?

File: tests/melt_compaction_elasticity.prm [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+# Following Keller, May & Kaus (2013, eq. 52), K_phi is porosity-dependent,
+# K_phi = K_0 * phi^(-1/2), with K_0 the reference pore modulus. Here the base
+# shear viscosity is 1e20 Pa s, the shear-to-bulk ratio is 1, phi_0 = 0.05 and
+# phi = 0.1, so xi = 1e20 * 0.05/0.1 = 5e19 Pa s. With K_0 = 1e9 Pa the initial
+# elastic pore modulus is K_phi = 1e9 * 0.1^(-1/2) ~= 3.16e9 Pa, so
+# tau = xi / K_phi = 5e19 / 3.16e9 ~= 1.58e10 s ~= 500 yr. (As the porosity
+# evolves during the run, K_phi and tau track it via eq. 52.) The point-values
+# postprocessor records the compaction pressure over time so this exponential
+# relaxation is captured in the reference output; a purely viscous run (large
+# default K_0) instead should yield p_c ~ 0 during the hold.
```
Comment (naliboff) [inline tests/melt_compaction_elasticity.prm]: Very cool test! 

A few thoughts:
1. Should this be moved to the benchmarks folder, which would allow adding plotting scripts etc for the analytical versus calculated solution?
2. At the end of the PRM description, could you provide an analytical value for p_c and the final predicted value from the test results?

File: doc/modules/changes/20260714_stoner (hunk 1 above)
Comment (ryanstoner1) [reply doc/modules/changes/20260714_stoner]: okay, updated now

File: source/simulator/melt.cc (hunk 3 above)
Comment (ryanstoner1) [reply source/simulator/melt.cc]: p_c_scale is being invoked earlier in melt.cc and not in the assemblers. I'm not sure if there was an additional guard there, but it seems like p_c_scale is the main one.

**Review by naliboff — COMMENTED** [review]
@ryanstoner1 - A few more comments and suggestions after going through the new tests cases and failing tests.

1. Would it be worth converting melt_compaction_elasticity to a benchmark?
2. It looks like the only test that fails is the new one you added melt_bending_beam_elasticity. Fantastic that none of the existing melt cases are affected!
3. I think it would be worth adding additional test cases (1-2) that demonstrated when porosity is zero one obtains the expected solution for non-melt material models. For example, the stress build-up benchmarks, etc.
4. The question of whether it will work to use fields for the elastic stresses and melt-related fields (operator splitting related). Is there an example test or benchmark that can be added to demonstrate this works? My concern is that the operator splitting requirements for elasticity (only 1  fixed step larger than maximum time step) is not compatible with the melt reactions.

File: source/simulator/melt.cc (hunk 2 above)
Comment (ryanstoner1) [reply source/simulator/melt.cc]: Patterns::Double(0.) already does what I was using this cutoff to do crudely (ensure >0 and non-inf).

Comment (ryanstoner1) [reply source/simulator/melt.cc]: So removed now.

File: benchmarks/viscoelastic_stress_build_up/viscoelastic_stress_build_up.prm [hunk 7]
```diff
@@ -73,8 +73,8 @@ subsection Geometry model
   set Model name = box
 
   subsection Box
-    set X repetitions = 50
-    set Y repetitions = 50
+    set X repetitions = 10
```
Comment (naliboff) [inline benchmarks/viscoelastic_stress_build_up/viscoelastic_stress_build_up.prm]: Was this change intentional?

File: source/material_model/reactive_fluid_transport.cc [hunk 8]
```diff
@@ -187,14 +192,24 @@ namespace aspect
...
-              // Scale the base model viscosity value based on the porosity.
+              // Scale the base model viscosity value based on the presence of fluid.
               for (unsigned int q=0; q<out.n_evaluation_points(); ++q)
                 {
                   const double porosity = std::max(in.composition[q][porosity_idx],0.0);
                   out.viscosities[q] *= (1.0 - porosity) * std::exp(- alpha_phi * porosity);
                 }
             }
 
+          // Keller et al. (2013) elastic force
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc]: Can you add the specific equation here for line 210?

File: source/material_model/reactive_fluid_transport.cc [hunk 9]
```diff
@@ -320,7 +337,8 @@ namespace aspect
                       else if (c == porosity_idx && this->get_timestep_number() > 0)
                         // Apply the volume fraction change to the porosity, which is a volume fraction
                         reaction_rate_out->reaction_rates[q][c] = porosity_change / reaction_time_step_size;
-                      else
+                      else if (composition_descriptions[c].type != CompositionalFieldDescription::stress)
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc:339]: I'm a bit confused by this addition, should this be `==` rather than` !=`? Likewise, why was this condition not caught with the existing `else` statement?

File: source/simulator/melt.cc [hunk 10]
```diff
@@ -542,32 +549,28 @@ namespace aspect
...
               if (scratch.rebuild_stokes_matrix)
                 for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
                   {
-                    data.local_matrix(i,j) += ( (eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
-                                                - (eta_two_thirds * (scratch.div_phi_u[i] * scratch.div_phi_u[j])
-                                                  )
-                                                - (pressure_scaling *
-                                                   scratch.div_phi_u[i] * scratch.phi_p[j])
+                    data.local_matrix(i, j) += ((eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
+                                                // The -2/3 eta (div u)(div v) deviatoric/compaction term only
```
Comment (naliboff) [inline source/simulator/melt.cc]: This is being added due to the current approach not correctly handling cell with no melt in incompressible models viscoelastic models. However, this the conditional statement below does not check if elasticity is being used in the model, but whether the model is incompressible and no melt is present? Can you explain a bit further about why this specific conditional statement was selected and whether it would affect prior cells in incompressible viscous models with no melt?

Comment (naliboff) [reply source/simulator/melt.cc]: @tjhei @jdannberg - It would be great to have your input here as well.

File: source/simulator/melt.cc (hunk 4 above)
Comment (naliboff) [reply source/simulator/melt.cc]: Summarizing a separate discussion (@ryanstoner1 @anne-glerum) - handling cases where fields are used for both elasticity and melt is possible, but would require quite a bit of additional work. This PR will be modified so that viscoelastic stresses can only be tracked with particles.

File: include/aspect/melt.h [hunk 11]
```diff
@@ -421,6 +424,20 @@ namespace aspect
...
+       * elastic pore modulus K_phi = K_0 * phi^(-1/2) (their eq. 52), which is
+       * the elastic modulus governing compaction (their eq. 32). A finite value
+       * turns the compaction channel into a Maxwell element: the effective
+       * compaction viscosity becomes xi*_ve = 1/(1/xi + 1/(K_phi dt)) and an
+       * elastic restoring source dP^o/(K_phi dt) (dP^o = old compaction pressure)
+       * is added to the compaction pressure equation. The default (a very large
+       * value) recovers purely viscous compaction, so existing models are
+       * unchanged.
+       */
+      double reference_pore_modulus;
```
Comment (naliboff) [inline include/aspect/melt.h]: Should the parameter go into the followup PR?

File: doc/modules/changes/20260714_stoner [hunk 12]
```diff
@@ -0,0 +1,15 @@
+New: Elasticity can now be combined with two-phase melt transport. The melt
```
Comment (naliboff) [inline doc/modules/changes/20260714_stoner:1]: The readme should be updated to remove the poroelasticity components, if the plan is still to add those in a follow-up PR.

File: tests/melt_compaction_elasticity.prm (hunk 6 above)
Comment (naliboff) [reply tests/melt_compaction_elasticity.prm]: Note - comments no longer apply to this PR, but hopefully helpful for a follow-up PR.

File: tests/melt_0percent_viscoelastic_bending_beam_particles.prm [hunk 13]
```diff
@@ -0,0 +1,97 @@
+# This test checks whether the viscoelastic bending beam benchmark
```
Comment (naliboff) [inline tests/melt_0percent_viscoelastic_bending_beam_particles.prm:1]: There are a number of existing test cases for the viscoelastic bending beam in tests folder, including one for particles (`viscoelastic_bending_beam_particles.prm`). How closely does this setup match the latter test case?  I think the ideal case would be to make the two setups identical (aside from the material model and melt-related changes) so the test output can be compared directly.

File: tests/melt_0percent_viscoelastic_bending_beam_particles.prm [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+subsection Initial composition model
+  subsection Function
+    set Function expression = 0; 0; 0; 0; 0; 0; if (x<=4.5e3 && y>=2.5e3 && y<=3.0e3, 1, 0); 0; 0
+  end
+end
+
+subsection Discretization
+  set Use discontinuous composition discretization = true, true, true, true, true, true, false, false, false
+  subsection Stabilization parameters
```
Comment (naliboff) [inline tests/melt_0percent_viscoelastic_bending_beam_particles.prm:57]: As none of the fields used the limiter, could this section be removed?

File: tests/melt_10percent_viscoelastic_bending_beam_particles.prm [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+subsection Initial composition model
+  subsection Function
+    set Function expression = 0; 0; 0; 0; 0; 0; if (x<=4.5e3 && y>=2.5e3 && y<=3.0e3, 1, 0); 0.1; 0
+  end
+end
+
+subsection Discretization
+  set Use discontinuous composition discretization = true, true, true, true, true, true, false, false, false
+  subsection Stabilization parameters
+    set Use limiter for discontinuous composition solution = false, false, false, false, false, false, false, false, false
```
Comment (naliboff) [inline tests/melt_10percent_viscoelastic_bending_beam_particles.prm:52]: Remove this section?

File: tests/melt_10percent_viscoelastic_bending_beam_particles.prm [hunk 16]
```diff
@@ -0,0 +1,98 @@
+# This test checks whether the viscoelastic bending beam benchmark
+# run successfully with two-phase flow. In particular, this benchmark is a good test to
```
Comment (naliboff) [inline tests/melt_10percent_viscoelastic_bending_beam_particles.prm:2]:
```suggestion
# run successfully with two-phase flow when there is non-zero porosity. In particular, this benchmark is a good test to
```

File: tests/melt_10percent_viscoelastic_bending_beam_particles.prm [hunk 17]
```diff
@@ -0,0 +1,98 @@
+# This test checks whether the viscoelastic bending beam benchmark
```
Comment (naliboff) [inline tests/melt_10percent_viscoelastic_bending_beam_particles.prm:1]: Can you update the documentation here to provide the following information:
1. Note whether this test is identical to melt_viscoelastic_bending_beam aside from the non-zero porosity and bound fluid content
2. Describe how the results change between these two tests.

File: tests/melt_0percent_viscoelastic_relaxation_particles.prm [hunk 18]
```diff
@@ -0,0 +1,205 @@
+# This setup models the relaxation of elastic stresses in the absence
+# of drivers.
+# The analytical solution is:
+#     \tau_{xx} = \tau_{xx_{t0}} \exp(-t \frac{\mu}{\eta_{viscous}}),
+# where \tau_{xx} is the first component of the stored stress tensor,
+# \mu is the shear modulus and \eta_{viscous} the viscous viscosity.
```
Comment (naliboff) [inline tests/melt_0percent_viscoelastic_relaxation_particles.prm:7]: Can you add a note confirming that the test result reproduces the analytical solution and what the specific values are? It would also be helpful to provide a short summary of the analytical solution for the values used in this PRM, so that they can be easily compared against the test results.

File: tests/melt_0percent_viscoelastic_relaxation_particles.prm [hunk 19]
```diff
@@ -0,0 +1,205 @@
+# This setup models the relaxation of elastic stresses in the absence
```
Comment (naliboff) [inline tests/melt_0percent_viscoelastic_relaxation_particles.prm:1]: Would it make sense to derive this PRM from the existing benchmark `viscoelastic_relaxation_particles.prm`? As no test exists for that specific benchmark case, it would be good to also a test for it using the regular viscoelastic material that can be compared directly against this new test case.

**Review by naliboff — CHANGES_REQUESTED** [review]
@ryanstoner1 - Thank you for the continued work on this PR.

Going forward, I think the next steps will primarily be:
1. Removing portions of the PR related to poroelasticity
2. Cleaning up and modifying the existing test cases
3. Adding a few additional test cases. Specifically, I think it would be good to add test cases for the `viscoelastic_stress_build_up_particles.prm`, `viscoelastic_stress_build_up_yield_particles.prm,` and `kaus_2010_extension.prm` benchmark cases.
4. Following on point 3, I wonder if the safest practice would be to actually add PRMs in the benchmarks folder for all of the new models using the melt interface for the zero porosity case. Thoughts on this @anne-glerum?

**Comment by anne-glerum** [comment]
> 4\. Following on point 3, I wonder if the safest practice would be to actually add PRMs in the benchmarks folder for all of the new models using the melt interface for the zero porosity case. Thoughts on this @anne-glerum?

Those prms would (should) reproduce the exact same results as the prms already in the benchmark folder, right? I'm not sure it's necessary to add them to the benchmark folder then. We also don't provide two prm files for the current two material models with elasticity that could produce them. It would be good to derive the test prms from the benchmark prms (with updated material model settings).

**Comment by ryanstoner1** [comment]
Yes, the prms are built to reproduce the same results as the benchmarks.

File: doc/modules/changes/20260714_stoner [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+Stokes assembler adds the stored elastic (deviatoric) shear stress as a force
+on the solid momentum equation (after Keller et al. 2013). Setting porosity to
+zero recovers purely viscous behavior predicted by single-phase flow (i.e. in
+cells without melt the melt momentum operator reduces exactly to the
+incompressible Stokes equation). Because elasticity requires the stress fields
+ to use a discontinuous discretization, melt transport now allows discontinuous
```
Comment (naliboff) [inline doc/modules/changes/20260714_stoner]:
```suggestion
to use a discontinuous discretization, melt transport now allows discontinuous
```

File: source/material_model/reactive_fluid_transport.cc [hunk 21]
```diff
@@ -187,14 +192,22 @@ namespace aspect
           // Modify the viscosity from the base model based on the presence of fluid.
           if (in.requests_property(MaterialProperties::viscosity))
             {
-              // Scale the base model viscosity value based on the porosity.
+              // check if shear elasticity present
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc]:
```suggestion
              // check if the model includes viscoelasticity
```

File: source/material_model/reactive_fluid_transport.cc [hunk 22]
```diff
@@ -320,7 +335,10 @@ namespace aspect
                       else if (c == porosity_idx && this->get_timestep_number() > 0)
                         // Apply the volume fraction change to the porosity, which is a volume fraction
                         reaction_rate_out->reaction_rates[q][c] = porosity_change / reaction_time_step_size;
-                      else
+                      else if (composition_descriptions[c].type != CompositionalFieldDescription::stress)
+                        // Previously zeroed all non-porosity or bound fluid reaction rates
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc]: Set the reaction rate to zero for all other fields with the exception of elastic stresses, where
the reaction rates were previously set in the base model and should not be overwritten.

File: source/simulator/solver_schemes.cc [hunk 23]
```diff
@@ -396,8 +396,21 @@ namespace aspect
         computing_timer.leave_subsection("Interpolate prescribed composition");
       }
 
+    // Enable interpolation of stresses on DG even if other particle-advected
```
Comment (naliboff) [inline source/simulator/solver_schemes.cc]: Update this documentation to describe more generally what is happening below. It should also be reviewed by someone with more knowledge of the fundamental particle/field mechanics (@gassmoeller?)

File: tests/melt_0percent_viscoelastic_bending_beam_particles.prm (hunk 13 above)
Comment (naliboff) [reply tests/melt_0percent_viscoelastic_bending_beam_particles.prm:1]: Follow-up - please add a note confirming that this test gives the exact same results as running the model without the melt interface if the solvers are switched to AMG.

File: tests/melt_0percent_viscoelastic_bending_beam_particles.prm (hunk 14 above)
Comment (naliboff) [reply tests/melt_0percent_viscoelastic_bending_beam_particles.prm:57]:
```suggestion
  subsection Stabilization parameters
  # While the default for 'set Use limiter for discontinuous composition solution' is 'false', it still 
  # needs to be set here for all fields as the test builds on a benchmark with a different number
  # of compositional fields. Not specifying these parameters for the new number of fields will
  # result in the model crashing.
```

File: tests/melt_10percent_viscoelastic_bending_beam_particles.prm (hunk 15 above)
Comment (naliboff) [inline tests/melt_10percent_viscoelastic_bending_beam_particles.prm:52]:
```suggestion
  # While the default for 'set Use limiter for discontinuous composition solution' is 'false', it still 
  # needs to be set here for all fields as the test builds on a benchmark with a different number
  # of compositional fields. Not specifying these parameters for the new number of fields will
  # result in the model crashing.
    set Use limiter for discontinuous composition solution = false, false, false, false, false, false, false, false, false
```

**Review by naliboff — COMMENTED** [review]
@ryanstoner1 - Per the live discussion, this is getting close!

Last key remaining steps:
1. Add a test that has some initial porosity present to an existing stress relaxation (or build-up) test case.
2. Have others review keys pieces of code in `melt.cc` and `solver_schemes.cc`

File: tests/melt_viscoelastic_bending_beam_particles.prm [hunk 24]
```diff
@@ -0,0 +1,108 @@
+# This test checks whether the viscoelastic bending beam benchmark
+# run successfully with two-phase flow. In particular, this benchmark is a good test to
```
Comment (naliboff) [inline tests/melt_viscoelastic_bending_beam_particles.prm]:
```suggestion
# runs successfully with two-phase flow. In particular, this benchmark is a good test to
```

File: doc/modules/changes/20260714_stoner [hunk 25]
```diff
@@ -0,0 +1,13 @@
+New: Elasticity can now be combined with two-phase melt transport. The melt
+Stokes assembler adds the stored elastic (deviatoric) shear stress as a force
+on the solid momentum equation (after Keller et al. 2013). Setting porosity to
+zero recovers purely viscous behavior predicted by single-phase flow (i.e. in
+cells without melt the melt momentum operator reduces exactly to the
+incompressible Stokes equation). Because elasticity requires the stress fields
+to use a discontinuous discretization, melt transport now allows discontinuous
+ elements for those stress fields (while other compositional fields and the
```
Comment (naliboff) [inline doc/modules/changes/20260714_stoner]:
```suggestion
elements for those stress fields (while other compositional fields and the
```

**Review by naliboff — COMMENTED** [review]
@ryanstoner1 - Thank you for adding the additional tests! I think this is nearly good from my side (pending rebase with #7219 once it is merged)

File: tests/melt_10percent_alpha0_viscoelastic_bending_beam_particles.prm
Comment (naliboff) [inline tests/melt_10percent_alpha0_viscoelastic_bending_beam_particles.prm:1]: @ryanstoner1 - Would it be good to rename this file to be consistent with the other test `melt_10percent_viscoelastic_relaxation_particles` that has 10% initial porosity? 

Likewise, would it be worth adding `weakening-factor` or similar to the name given this varies between tests?

This could be taken further and adding `0percent` to the models with no porosity. Al

Some something like `melt_0percent-porosity_1weakening_`?

I am also fine with all of this being updated in a follow-up PR.

**Review by naliboff — APPROVED** [review]
@ryanstoner1 - Thank you for the additional tests and rebasing! I had one additional comment about the naming scheme for the new tests, but this can be addressed in a follow-up PR. As such, I've approved the PR and will let you decide how and when to handle unifying the test naming schemes.

@tjhei @jdannberg - Would you like to have a final look before merging?

File: doc/modules/changes/20260714_stoner [hunk 26]
```diff
@@ -0,0 +1,13 @@
+New: Elasticity can now be combined with two-phase melt transport. The melt
+Stokes assembler adds the stored elastic (deviatoric) shear stress as a force
+on the solid momentum equation (after Keller et al. 2013). Setting porosity to
+zero recovers purely viscous behavior predicted by single-phase flow (i.e. in
+cells without melt the melt momentum operator reduces exactly to the
+incompressible Stokes equation). Because elasticity requires the stress fields
+to use a discontinuous discretization, melt transport now allows discontinuous
+elements for those stress fields (while porosity and temperature must still be
+continuous). In addition, visco-plastic strain weakening and the elastic-stress
```
Comment (tjhei) [inline doc/modules/changes/20260714_stoner]:
```suggestion
incompressible Stokes equation). In addition, visco-plastic strain weakening and the elastic-stress
```

this was done separately, right?

**Review by tjhei — APPROVED** [review]

**Comment by naliboff** [comment]
@ryanstoner1 - Thank you for updating the test naming schemes, this is incredibly helpful for an organizational perspective!
